rfc-087-dealing-with-errors.md

```diff
@@ -0,0 +1,105 @@
+# Dealing with errors
+
+## Summary
+
+This describes how we treat errors on GOV.UK
+
+## Problem
+
+We've recently migrated to a new error tracking service, Sentry. This provides us an opportunity to rethink how we treat errors.
+
+## Proposal
+
+There are 2 principles:
+
+### 1. When something goes wrong, we should be notified
+
+Applications should report exceptions to Sentry. Applications must not swallow errors.
+
+### 2. Applications should not error
+
+The goal of GOV.UK is that applications should not error. When something goes wrong it should be fixed.
+
+## Classifying errors
+
+### Bug
+
+A code change makes the application crash.
+
+Desired behaviour: error is sent to Sentry, developers are notified and fix the error. Developers mark the error in Sentry as `Resolved`. This means a recurrence of the error will alert developers again.
+
+### Intermittent errors without user impact
+
+Frontend applications often see timeouts when talking to the content-store.
+
+There's no or little user impact because the request will be answered by the caching layer.
+
+Example: <https://sentry.io/govuk/app-finder-frontend/issues/352985400>
+
+Desired behaviour: error is not sent to Sentry. Instead, we rely on Smokey and Icinga checks to make sure we the site functions.
+
+### Intermittent errors with user impact
```

Comment:
Availability errors like timeouts can cause problems for users, but I don't think they are very actionable in sentry. In particular, for Rummager, it's often hard to distinguish between timeouts in our API (which can interfere with rendering) and timeouts that take place within different contexts like rake tasks or sidekiq workers.

We should absolutely be tracking when this happens (for example by storing metrics in graphite) but raising sentry errors doesn't seem helpful unless there is an obvious way to intervene.

Reply:
I think that responding to failure of dependent services is a concern that should be handled in shared libraries, rather than applications themselves, because it's tricky to get right and we should have common approaches to handling it. Have a shared solution for in band HTTP requests, a shared solution for background workers etc.

Reply:
Updated in https://github.com/alphagov/govuk-rfcs/pull/87/commits/494975f9ce8e43e89b028413cd35c267f3e9d268. Do you agree?